tests/test_json_schema.py

```diff
@@ -6257,6 +6257,22 @@ class Model(BaseModel):
     }
 
 
+def test_plain_serializer_does_not_apply_json_unless_none() -> None:
+    """Test plain serializers aren't used to compute the JSON Schema default if mode is 'json-unless-none'
+    and default value is `None`."""
+
+    class Model(BaseModel):
+        custom_decimal: Annotated[
+            Decimal | None, PlainSerializer(lambda x: float(x), when_used='json-unless-none', return_type=float)
+        ] = None
+
+    assert Model.model_json_schema(mode='serialization') == {
+        'properties': {'custom_decimal': {'default': None, 'title': 'Custom Decimal', 'type': 'number'}},
+        'title': 'Model',
+        'type': 'object',
+    }
```

Comment:
This revealed another issue with the JSON Schema: `type` should be `'anyOf': [{'type': 'number'}, {'type': 'null'}]}`.

When the JSON Schema for our field is created, `schema` is of type `default`, and `schema['schema']` is of type `nullable`:

https://github.com/pydantic/pydantic/blob/0c97049bd40ac95cce7bbebad8369b9426af4b09/pydantic/json_schema.py#L1022-L1031

However, `self.generate_inner(schema['schema']`) will not go through `nullable_schema` as we would expect. As there's a `serialization` key, `ser_schema` is being used:

https://github.com/pydantic/pydantic/blob/0c97049bd40ac95cce7bbebad8369b9426af4b09/pydantic/json_schema.py#L492-L494

This makes sense, however, the issue is `ser_schema` kind of assumes `when_used` is `always`. I'll see if I can fix this in a follow up PR.